File: frontend/omgnice_ecommerce_app/lib/features/promotion/data/models/promotion_model.dart
import 'package:omgnice_ecommerce_app/features/promotion/domain/entities/promotion.dart';

class PromotionModel extends PromotionEntity {
  PromotionModel({
    int? id,
    String? code,
    String? title,
    String? description,
    String? discountType,
    double? discountValue,
    double? maxDiscountValue,
    double? minOrderValue,
    String? appliesTo,
    int? productId,
    int? categoryId,
    DateTime? startDate, 
    DateTime? endDate,
    int? usageLimit,
    int? usedCount,
    bool? isActive,
    bool? isExclusive,
  }) : super(
          id: id,
          code: code,
          title: title,
          description: description,
          discountType: discountType,
          discountValue: discountValue,
          maxDiscountValue: maxDiscountValue,
          minOrderValue: minOrderValue,
          appliesTo: appliesTo,
          productId: productId,
          categoryId: categoryId,
          startDate: startDate,
          endDate: endDate,
          usageLimit: usageLimit,
          usedCount: usedCount,
          isActive: isActive,
          isExclusive: isExclusive,
        );

  factory PromotionModel.fromJson(Map<String, dynamic> json) {
    return PromotionModel(
      id: json['id'],
      code: json['code'],
      title: json['title'],
      description: json['description'],
      discountType: json['discount_type'],
      discountValue: (json['discount_value'] as num?)?.toDouble(),
      maxDiscountValue: (json['max_discount_value'] as num?)?.toDouble(),
      minOrderValue: (json['min_order_value'] as num?)?.toDouble(),
      appliesTo: json['applies_to'],
      productId: json['product_id'],
      categoryId: json['category_id'],
      startDate: json['start_date'] != null ? DateTime.parse(json['start_date']) : null,
      endDate: json['end_date'] != null ? DateTime.parse(json['end_date']) : null,
      usageLimit: json['usage_limit'],
      usedCount: json['used_count'],
      isActive: json['is_active'],
      isExclusive: json['is_exclusive'] ?? false, // Default to false if not provided
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'title': title,
      'description': description,
      'discount_type': discountType,
      'discount_value': discountValue,
      'max_discount_value': maxDiscountValue,
      'min_order_value': minOrderValue,
      'applies_to': appliesTo,
      'product_id': productId,
      'category_id': categoryId,
      'start_date': startDate?.toIso8601String(),
      'end_date': endDate?.toIso8601String(),
      'usage_limit': usageLimit,
      'used_count': usedCount,
      'is_active': isActive,
    };
  }

  @override
  String toString() {
    return 'PromotionModel(id: $id, code: $code, title: $title, description: $description, discountValue: $discountValue)';
  }
}

File: backend/models/promotion.model.js
module.exports = (sequelize, DataTypes) => {
    const promotion = sequelize.define('Promotion', {
      id: {
        type: DataTypes.INTEGER,
        autoIncrement: true,
        allowNull: false,
        primaryKey: true,
      },
  
      code: {
        type: DataTypes.STRING,
        allowNull: true,
        unique: true, // mã giảm giá duy nhất
      },
  
      title: {
        type: DataTypes.STRING,
        allowNull: false,
      },
  
      description: {
        type: DataTypes.TEXT,
        allowNull: true,
      },
  
      discount_type: {
        type: DataTypes.ENUM('PERCENTAGE', 'FIXED'),
        allowNull: false,
      },
  
      discount_value: {
        type: DataTypes.FLOAT,
        allowNull: false,
      },
  
      max_discount_value: {
        type: DataTypes.FLOAT,
        allowNull: true, // chỉ áp dụng nếu là dạng phần trăm
      },
  
      min_order_value: {
        type: DataTypes.FLOAT,
        allowNull: true, // đơn tối thiểu để áp dụng
      },
  
      applies_to: {
        type: DataTypes.ENUM('ALL', 'PRODUCT', 'CATEGORY'),
        defaultValue: 'ALL',
      },
  
      product_id: {
        type: DataTypes.INTEGER,
        allowNull: true,
      },
  
      category_id: {
        type: DataTypes.INTEGER,
        allowNull: true,
      },
  
      start_date: {
        type: DataTypes.DATE,
        allowNull: false,
      },
  
      end_date: {
        type: DataTypes.DATE,
        allowNull: false,
      },
  
      usage_limit: {
        type: DataTypes.INTEGER,
        allowNull: true, // tổng số lượt dùng
      },
  
      used_count: {
        type: DataTypes.INTEGER,
        defaultValue: 0,
      },
  
      is_active: {
        type: DataTypes.BOOLEAN,
        defaultValue: true,
      },
      is_exclusive: {
  type: DataTypes.BOOLEAN,
  defaultValue: false,
  comment: 'Chỉ dành riêng cho khách hàng đặc biệt, không công khai',
},

    }, {
      tableName: 'promotions',
      timestamps: false,
    });
  
    return promotion;
  };
  